File: BFS & DFS implementation/List.c
/****************************************************************************************
*  List.c
*  Name: Tung-Lin Lee
*  This file will implement the List ADT operation
*****************************************************************************************/
#include<stdio.h>
#include<stdlib.h>

#include"List.h"

// structs


// private NodeObj type
typedef struct NodeObj{
    int data;
    struct NodeObj* next;
    struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj typedef
typedef struct ListObj{
    Node front;
    Node back;
    Node cursor;
    int length;
    int index;
} ListObj;

// newList()
// Constructor of the List
List newList(){
    List L = malloc(sizeof(ListObj));
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->index = -1;
    return L;
}

// freeList()
// Destructor of the List
void freeList(List* pL){
    // free the memory of the nodes
    clear(*pL);

    // free the heap memory
    (*pL)->back = NULL;
    (*pL)->front = NULL;
    (*pL)->cursor = NULL;
    (*pL)->length = 0;

    free(*pL);
    *pL = NULL;
}

// Access functions

// length()
// return the length of the List
int length(List L){
    if(L == NULL){
//        printf("This List is empty\n");
        return -1;
    }
    return L->length;
}

// index()
// return the index of the cursor
int index(List L){
    if(L->cursor != NULL){
        return L->index;
    }
     else{
        return -1;
    }
}

// front()
// Pre: length() > 0
// return the element of the 1st element in the List
int front(List L){
    if(length(L) > 0){
        return L->front->data;
    }
    else{
        return -1;
    }
}

// back()
// length() > 0
// return the last element in the List
int back(List L){
    if(length(L) > 0){
        return L->back->data;
    }
    else{
        return -1;
    }
}

// get()
// Pre: length>0, index>=0
// return the element of the cursor
int get(List L){
    if(length(L) > 0 && index(L) >= 0){
//        printf("inside stack: %d\n",L->cursor->data);
        return L->cursor->data;
    }
    else{
        return -1;
    }
}

// equal()
// if two list are equal, return 1
// return 0 otherwise
int equals(List A, List B){

    // if both List are empty, return 1
    if(A->front == NULL && B->front == NULL){
        return 1;
    }

    int retval = 1;
    Node walkerA = A->front;
    Node walkerB = B->front;
    while(walkerA != NULL && walkerB != NULL){
        if(walkerA->data != walkerB->data){
            retval = 0;
            break;
        }
        walkerA = walkerA->next;
        walkerB = walkerB->next;
    }

    if((walkerA == NULL && walkerB != NULL) || (walkerA != NULL && walkerB == NULL)) {
        retval = 0;
    }
    return retval;
}

// Manipulation procedures
// clear()
// clear the nodes in the List, and set all the attribute to original state
void clear(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    Node tempNode = L->front;
    Node nodePtr;
    while(tempNode != NULL){
        nodePtr = tempNode;
        tempNode = tempNode->next; //
        free(nodePtr);
    }

    // initialize the attributes to a empty list
    L->front = NULL;
    L->back = NULL;
    L->cursor = NULL;
    L->length = 0;
    L->index = -1;
}

// moveFront()
// move the cursor to the front
void moveFront(List L){
    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(length(L) != 0){
        L->index = 0;
        L->cursor = L->front;
    }
}

// moveBack()
// move the cursor to the back
void moveBack(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(length(L) != 0){
        L->index = length(L)-1;
        L->cursor = L->back;
    }
}

// movePrev()
// move the cursor one step before
void movePrev(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(index(L) != -1 && L->cursor != L->front){
        L->cursor = L->cursor->prev;
        L->index--;
    }
    else if (index(L) != -1 && L->cursor == L->front){
        L->cursor = NULL;
        L->index = -1;
    }
}

// moveNext()
// move the cursor one step after
void moveNext(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(index(L) != -1 && L->cursor != L->back){
        L->cursor = L->cursor->next;
        L->index++;
    }
    else if(index(L) != -1 && L->cursor == L->back){
        L->cursor = NULL;
        L->index = -1;
    }
}

// prepend()
// insert the element in the end of the LIst
void prepend(List L, int data){

    Node newNode = malloc(sizeof(NodeObj));
    newNode->data = data;
    newNode->next= NULL;
    newNode->prev = NULL;

    if(length(L) == 0){
        L->front = newNode;
        L->back = L->front;
        newNode->next = NULL;
        newNode->prev = NULL;
    }
    else{
        newNode->next = L->front;
        L->front->prev = newNode;
        L->front = newNode;
    }
    L->length++;
    if(index(L)!= -1){
        L->index++;
    }
}

//append()
// insert the element in the beginning of the List
void append(List L, int data){

    Node newNode = malloc(sizeof(NodeObj));
    newNode->data = data;
    newNode->next= NULL;
    newNode->prev = NULL;

    if(length(L) == 0){
        L->front = newNode;
        L->back = L->front;
        newNode->next = NULL;
        newNode->prev = NULL;
    }
    else{
        newNode->prev = L->back;
        L->back->next = newNode;
        L->back = newNode;
        newNode->next = NULL;
    }
    L->length++;
}

// insertBefore()
// Pre: index>=0
// insert the element before the cursor
void insertBefore(List L, int data){

    Node newNode = malloc(sizeof(NodeObj));
    newNode->data = data;
    newNode->next= NULL;
    newNode->prev = NULL;

    if(L->cursor->prev == L->front){
        newNode->next = L->cursor;
        L->cursor->prev = newNode;
        L->front->next = newNode;
        newNode->prev = L->front;
    }
    else if(L->cursor == L->front){
        L->front->prev = newNode;
        newNode->next = L->front;
        L->front = newNode;
    }
    else{
        Node* oriBeforeCur = L->cursor->prev;
        L->cursor->prev->next = newNode;
        newNode->next = L->cursor;
        L->cursor->prev = newNode;
        newNode->prev = oriBeforeCur;
    }

    L->index++;
    L->length++;
}


// insertAfter()
// Pre: index >=0
// insert the element after the cursor
void insertAfter(List L, int data){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    Node newNode = malloc(sizeof(NodeObj));
    newNode->data = data;
    newNode->next= NULL;
    newNode->prev = NULL;

    if(L->cursor == L->back){
        L->back->next = newNode;
        newNode->prev = L->back;
        L->back = newNode;
        newNode->next = NULL;
    }
    else{
        Node oriAftCur = L->cursor->next;
        L->cursor->next = newNode;
        newNode->prev = L->cursor;
        oriAftCur->prev = newNode;
        newNode->next = oriAftCur;
    }
    L->length++;
}


//deleteFront()
// Pre: length > 0
// delete the element in the 1st position
void deleteFront(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(length(L) > 0){

        if(length(L) == 1){
            clear(L);
        }
        else{
            if(L->cursor == L->front){
            L->cursor = NULL;
            L->index = -1;
            }
            Node newFront = L->front->next;
            free(L->front);
            L->front = newFront;
            L->front->prev = NULL;
            L->length--;
        }

        if(index(L) != -1){
            L->index--;
        }
    }
}

// deleteBack()
// length >0
// delete the element in the last position
void deleteBack(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    if(length(L) > 0){
        if(L->back == L->cursor){
            L->cursor = NULL;
            L->index = -1;
        }
        if(length(L) == 1){
            clear(L);
        }
        else{
            Node newBack = L->back->prev;
            free(L->back);
            L->back = newBack;
            newBack->next = NULL;
            L->length--;
        }
    }
}


// delete()
// length()>0, index()>=0
// delete the element in the cursor's position
void delete(List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }
    if(length(L) == 1){
        clear(L);
    }
    else{
         if(length(L) >= 0 && index(L) >= 0){
            if(L->cursor != L->front && L->cursor != L->back){
                Node pre = L->cursor->prev;
                Node aft = L->cursor->next;
                pre->next = aft;
                aft->prev = pre;
            }
            if(L->cursor == L->front){
                if(length(L) == 2){
                    L->front = L->back;
                    L->back->prev = NULL;
                }
                else{
                    L->front = L->cursor->next;
                    L->front->prev = NULL;
                }
            }
            if(L->cursor == L->back){
                if(length(L) == 2){
                    L->back = L->front;
                    L->front->next = NULL;
                }
                else{
                    L->back = L->cursor->prev;
                    L->back->next = NULL;
                }
            }
        }
        free(L->cursor);
        L->cursor = NULL;
        L->index = -1;
        L->length--;
    }
}

//Other operations

// printList()
// print out the elements in the List
void printList(FILE* out, List L){

    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    Node walker = L->front;
    while(walker != NULL){
        fprintf(out, "%d ", walker->data);
        walker = walker->next;
    }
}

// copyList()
// copy the List to the other
List copyList(List L){
    if(L == NULL){
//        printf("This List is empty\n");
        exit(1);
    }

    List copyList = newList();
    Node oriWalker = L->front;
    while(oriWalker != NULL){
        append(copyList, oriWalker->data);
        oriWalker = oriWalker->next;
    }

    return copyList;
}

File: Adjacency Matrix/List.java
//-----------------------------------------------------------------------------
//List.java
//Name: Tung-Lin Lee
//This file will implement the List ADT operation
//-----------------------------------------------------------------------------


@SuppressWarnings("overrides")
public class List {

// private attributes
private int length;
private int index; // for the index of cursor
private Node front;
private Node rear;
private Node cursor;

// Inner class. it will create the node to store and link each element in the
// List
@SuppressWarnings("overrides")
private class Node {
	Object item;
	Node next;
	Node prev;
	
	@Override
	public boolean equals(Object x){
		boolean status = true;
		Node N;
		if( x instanceof Node ) {
			N = (Node)x;
			status = this.item.equals(N.item);
		}
		return status;
    }
}

/******************
 * Constructor
 *******************/

/**
 * List() Creates a new empty list.
 */
public List() {
	this.length = 0;
	this.index = -1;
	this.front = null;
	this.rear = null;
	this.cursor = null;
}

/****************************************************************
 * Access functions
 *****************************************************************/

/**
 * int length() this method will return the number of the element in the List
 * Pre: None
 */
public int length() {
	return this.length;
}

/**
 * int index() If cursor is defined, returns the index of the cursor element,
 * otherwise returns -1 Pre: None
 */
public int index() {
	if (cursor != null) {
		Node walker = front;
		int count = 0;
		while (walker != cursor) {
			walker = walker.next;
			count++;
		}
		return count;
	} else {
		return -1;
	}
}

/**
 * int front() Returns front element. if the List is empty, it will return -1
 * Pre: length()>0
 */
public Object front() {
	if (length() > 0) {
		return front.item;
	} else {
		return -1;
	}
}

/**
 * int back() Returns back element. if the List is empty, it will return -1 Pre:
 * length()>0
 */
public Object back() {
	if (length() > 0) {
		return rear.item;
	} else {
		return -1;
	}
}

/**
 * int get() Returns cursor element. if the cursor is undefined, it will return
 * -1 Pre: length()>0, index()>=0
 */
public Object get() {
	if (length() > 0 && index() >= 0) {
		return cursor.item;
	} else {
		return -1;
	}
}

/**
 * boolean equals(List) Returns true if and only if this List and L are the same
 * integer sequence. The states of the cursors in the two Lists are not used in
 * determining equality.
 */
@Override
public boolean equals(Object x) {
	boolean status = true;

	// if the length of the both lists are not same, return false	
	if(this.length != ((List)x).length) {
		return false;
	}
	else {
		Node ptr1 = front;
		Node ptr2 = ((List)x).front;
		while(ptr1 != null) {
			if(ptr1.equals(ptr2)) {
				ptr1 = ptr1.next;
				ptr2 = ptr2.next;
			}
			else {
				return false;
			}
		}
		return status;
	}
}

/**************************************
 * Manipulation procedures
 **************************************/

/**
 * void clear() Resets this List to its original empty state.
 */
public void clear() {
	
	// delete entire list
	front = null;
	rear = null;
	
	// initialize the attribute
	index = -1;
	length = 0;
	cursor = null;
}

/**
 * void moveFront() If List is non-empty, places the cursor under the front
 * element, otherwise does nothing.
 */
public void moveFront() {
	if (length() != 0) {
		cursor = front;
		index = 0;
	}
}

/**
 * void moveBack() If List is non-empty, places the cursor under the back
 * element, otherwise does nothing.
 */
public void moveBack() {
	if (length() != 0) {
		cursor = rear;
		index = length() - 1;
	}
}

/**
 * void movePrev() If cursor is defined and not at front, moves cursor one step
 * toward front of this List, if cursor is defined and at front, cursor becomes
 * undefined, if cursor is undefined does nothing.
 */
public void movePrev() {
	if (index != -1 && cursor != front) {
		cursor = cursor.prev;
		index--;
	} else if (index != -1 && cursor == front) {
		cursor = null;
		index = -1;
	}
}
/**
 * void moveNext() If cursor is defined and not at back, moves cursor one step
 * toward back of this List, if cursor is defined and at back, cursor becomes
 * undefined, if cursor is undefined does nothing.
 */
public void moveNext() {
	if (index != -1 && cursor != rear) {
		cursor = cursor.next;
		index++;
	} else if (index != -1 && cursor == rear) {
		cursor = null;
		index = -1;
	}
}

/**
 * void prepend(int) Insert new element into this List. If List is non-empty,
 * insertion takes place before front element.
 */
public void prepend(Object data) {

	// insert the data to the new node
	Node newNode = new Node();
	newNode.item = data;

	// if the List is empty, make front and back point to this node
	if (length() == 0) {
		front = newNode;
		rear = front;
		newNode.next = null;
		newNode.prev = null;
	} else {
		newNode.next = front;
		front.prev = newNode;
		front = newNode;
	}
	length++;

	// if the new node is inserted before the front and replace the front position,
	// change the old front to the back
	if (length() == 2) {
		newNode.next = rear;
	}
}

/**
 * void append(int) Insert new element into this List. If List is non-empty,
 * insertion takes place after back element.
 */
public void append(Object data) {

	// insert the data into the new node
	Node newNode = new Node();
	newNode.item = data;

	// if the List is empty, make front and back point to this node
	if (length() == 0) {
		front = newNode;
		rear = front;
		newNode.next = null;
		newNode.prev = null;
	} else {
		newNode.prev = rear;
		rear.next = newNode;
		rear = newNode;
	}
	length++;
}

/**
 * void insertBefore(int) Insert new element before cursor. Pre: length()>0,
 * index()>=0
 */
public void insertBefore(Object data) {

	// insert the new element into the new node
	Node newNode = new Node();
	newNode.item = data;

	// case 1: when cursor's previous element is front
	if (cursor.prev == front) {
		newNode.next = cursor;
		cursor.prev = newNode;
		front.next = newNode;
		newNode.prev = front;
	}
	// case 2: when cursor is in the front
	else if (cursor == front) {
		front.prev = newNode;
		newNode.next = front;
		front = newNode;
	}
	// case 3: when cursor's previous element is normal node
	else {
		Node oriBeforeCur = cursor.prev;
		cursor.prev.next = newNode;
		newNode.next = cursor;
		cursor.prev = newNode;
		newNode.prev = oriBeforeCur;
	}

	// update the List attribute
	index++; // increment by 1 since new element insert before cursor
	length++;
}

/**
 * void insertAfter(int) Inserts new element after cursor. Pre: length()>0,
 * index()>=0
 */
public void insertAfter(Object data) {
	// insert the new element into the new node
	Node newNode = new Node();
	newNode.item = data;

	// if insert after the back, update the back
	if (cursor == rear) {
		rear.next = newNode;
		newNode.prev = rear;
		rear = newNode;
	} else {
		Node oriAftCur = cursor.next; // oriAftCur stand for Originally After Cursor
		cursor.next = newNode;
		newNode.prev = cursor;
		oriAftCur.prev = newNode;
		newNode.next = oriAftCur;
	}

	// update the List attribute
	length++;
}

/**
 * void deleteFront() Deletes the front element. Pre: length()>0
 */
public void deleteFront() {
	if (length() > 0) {
		if (length == 1) { // if there's only one node
			front = null;
			rear = null;
			length--;
		} else {
			if (cursor == front) { // if cursor is in the front position but it needs to delete
				cursor = null; // set the cursor to null
				index = -1;
			}
			Node newFront = front.next;
			front = null;
			front = newFront;
			front.prev = null;
			length--;
		}
	}
}

/**
 * void deleteBack() Deletes the back element. Pre: length()>0
 */
public void deleteBack() {
	if (length() > 0) {
		if (length == 1) { // if there's only one node
			front = null;
			rear = null;
			length--;
		} else {
			if (cursor == rear) { // if the cursor is in back position
				cursor = null; // set the cursor to undefined
				index = -1;
			}
			Node newBack = rear.prev;
			rear = null;
			rear = newBack;
			newBack.next = null;
			length--;
		}
	}
}

/**
 * void delete() Deletes cursor element, making cursor undefined. Pre:
 * length()>0, index()>=0
 */
public void delete() {

	if (length() >= 0 && index() >= 0) {
		// case 1: if the cursor is not in front and back
		if (cursor != front && cursor != rear) {
			Node pre = cursor.prev;
			Node aft = cursor.next;
			pre.next = aft;
			aft.prev = pre;
		}
		// case 2: if the cursor is in the front position
		else if (cursor == front) {
			front = cursor.next;
			cursor.prev = null;
		}
		// case 3: if the cursor is in the back position
		else if (cursor == rear) {
			if (length == 2) { // if there is only two node
				rear = front;
				front.next = null;
			} else {
				rear = cursor.prev;
				rear.next = null;
			}
		}

		// set the cursor to undefined and decrement the length by 1
		cursor = null;
		index = -1;
		length--;
	}
}

/***************************************
 * other methods
 ***************************************/

/**
 * String toString() Overrides Object's toString method. Returns a String
 * representation of this List consisting of a space separated sequence of
 * integers, with front on left.
 */
@Override
public String toString() {
	Node nodeWalker = front;
	String retVal = "";
	while (nodeWalker != null) {
		retVal = retVal.concat(" " + String.valueOf(nodeWalker.item));
		nodeWalker = nodeWalker.next;
	}
	retVal = retVal.trim(); // remove the white space in the end of the string
	return retVal;
} 


}
